railties/lib/rails/generators/rails/sessions/templates/controllers/concerns/authentication.rb

```diff
@@ -0,0 +1,64 @@
+module Authentication
+  extend ActiveSupport::Concern
+
+  included do
+    before_action :require_authentication
+    helper_method :authenticated?
+  end
+
+  class_methods do
+    def allow_unauthenticated_access(**options)
+      skip_before_action :require_authentication, **options
+    end
+  end
+
+  private
+    def authenticated?
+      Current.session.present?
+    end
+
+    def require_authentication
+      resume_session || request_authentication
+    end
+
+
+    def resume_session
+      if session = find_session_by_cookie
+        set_current_session session
+      end
+    end
+
+    def find_session_by_cookie
+      if token = cookies.signed[:session_token]
+        Session.find_by(token: token)
+      end
+    end
+
+
+    def request_authentication
+      session[:return_to_after_authenticating] = request.url
+      redirect_to new_session_url
+    end
+
+    def after_authentication_url
+      session.delete(:return_to_after_authenticating) || root_url
+    end
+
+
+    def start_new_session_for(user)
+      user.sessions.create!(user_agent: request.user_agent, ip_address: request.remote_ip).tap do |session|
+        set_current_session session
```

Comment:
Nitpick: The name `session` could be confused with request's `session` object. Should it be called `current_session` for clarity ?

Same for `set_current_session(session)`

Reply:
Yup, in fact authenticated? seems pointless with this implementation. There's 3 paths here

1) require_authentication && resume > authenticated? is always true
2) require_authentication && request > it redirects to login
3) skip_authentication > authenticated? is always false

Seems like resume session needs to be called regardless, and request_authentication is the skippable part

In my setup, which is very similar to this I have 3 before action type helpers

require_authentication (resume || redirect)
allow_authentication (resume)
despise_authentication (resume && redirect)

require covers most of the cases, allow is useful in some areas where it's not necessary but we don't forbid, and despise will kick you out if your signed in. Think invitation/signup type controllers.